### PR TITLE
refactor(tools): separate IP-related functions from tool.utils

### PR DESCRIPTION
**Pull request by Water-Melon**
### Summary

Move IP-related variables and functions from `kong/tools/utils.lua` to `kong/tools/ip.lua`.

### Issue reference

KAG-2958

### CODE REVIEW

File: kong/tools/ip.lua [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+local _M = {}
+
+
+local ipv4_prefixes = {}
+for i = 0, 32 do
+  ipv4_prefixes[tostring(i)] = i
+end
+
+local ipv6_prefixes = {}
```
Comment (chronolaw) [inline kong/tools/ip.lua:21]: Two blank lines between functions.

File: kong/tools/ip.lua [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    -- has port number
+    a,b,c,d,port = address:match("^(%d%d?%d?)%.(%d%d?%d?)%.(%d%d?%d?)%.(%d%d?%d?):(%d+)$")
+  else
+    -- without port number
+    a,b,c,d,port = address:match("^(%d%d?%d?)%.(%d%d?%d?)%.(%d%d?%d?)%.(%d%d?%d?)$")
+  end
+  if not a then
+    return nil, "invalid ipv4 address: " .. address
+  end
+  a,b,c,d = tonumber(a), tonumber(b), tonumber(c), tonumber(d)
```
Comment (chronolaw) [inline kong/tools/ip.lua:132]: localize `tonumber`

File: kong/tools/ip.lua [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  local match = name:match("^[%d%a%-%.%_]+$")
+  if match == nil then
+    return nil, "invalid hostname: " .. address
+  end
+
+  -- Reject prefix/trailing dashes and dots in each segment
+  -- notes:
+  --   - punycode allows prefixed dash, if the characters before the dash are escaped
+  --   - FQDN can end in dots
+  for index, segment in ipairs(split(name, ".")) do
```
Comment (chronolaw) [inline kong/tools/ip.lua:227]: localize `ipairs`

File: kong/tools/ip.lua [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+-- @param p1 address to format, either string with name/ip, table returned from `normalize_ip`, or from the `socket.url` library.
+-- @param p2 port (optional) if p1 is a table, then this port will be inserted if no port-field is in the table
+-- @return formatted address or nil+error
+-- @usage
+-- local addr, err = format_ip(normalize_ip("001.002.003.004:123"))  --> "1.2.3.4:123"
+-- local addr, err = format_ip(normalize_ip("::1"))                  --> "[0000:0000:0000:0000:0000:0000:0000:0001]"
+-- local addr, err = format_ip("::1", 80))                           --> "[::1]:80"
+-- local addr, err = format_ip(check_hostname("//bad .. name\\"))    --> nil, "invalid hostname: ... "
+function _M.format_host(p1, p2)
+  local t = type(p1)
```
Comment (chronolaw) [inline kong/tools/ip.lua:274]: localize `type`

File: kong/tools/ip.lua [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+function _M.normalize_ip(address)
+  local atype = _M.hostname_type(address)
+  local addr, port = verify_types[atype](address)
+  if not addr then
+    return nil, port
+  end
+  return {
+    type = atype,
+    host = addr,
+    port = port
```
Comment (chronolaw) [inline kong/tools/ip.lua:256]: need a comma at the end of line.

File: kong/tools/ip.lua [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+  return name, port
+end
+
+
+local verify_types = {
+  ipv4 = _M.normalize_ipv4,
+  ipv6 = _M.normalize_ipv6,
+  name = _M.check_hostname,
+}
+--- verifies and normalizes ip adresses and hostnames. Supports ipv4, ipv4:port, ipv6, [ipv6]:port, name, name:port.
```
Comment (chronolaw) [inline kong/tools/ip.lua:242]: two blank lines.

File: kong/tools/ip.lua (hunk 1 above)
Comment (Water-Melon) [reply kong/tools/ip.lua:25]: done

File: kong/tools/ip.lua (hunk 2 above)
Comment (Water-Melon) [reply kong/tools/ip.lua:147]: done

File: kong/tools/ip.lua (hunk 3 above)
Comment (Water-Melon) [reply kong/tools/ip.lua:242]: done

File: kong/tools/ip.lua (hunk 4 above)
Comment (Water-Melon) [reply kong/tools/ip.lua:291]: done

File: kong/tools/ip.lua (hunk 5 above)
Comment (Water-Melon) [reply kong/tools/ip.lua]: done

File: kong/tools/ip.lua (hunk 6 above)
Comment (Water-Melon) [reply kong/tools/ip.lua:259]: done

File: kong/tools/ip.lua [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+local gsub     = string.gsub
+local sub      = string.sub
+local fmt      = string.format
+local lower    = string.lower
+local find     = string.find
+local split    = pl_stringx.split
+local type     = type
+local ipairs   = ipairs
+local tonumber = tonumber
```
Comment (chronolaw) [inline kong/tools/ip.lua:13]: These functions are build-in functions, should be in front of `gsub`

Comment (Water-Melon) [reply kong/tools/ip.lua]: done

**Review by chronolaw — APPROVED** [review]

**Review by fffonion — APPROVED** [review]

**Pull request merged**